### PR TITLE
New ball placement play test

### LINKED ISSUE #2780: Validation for ball placement play
### Description of the task
Currently, `ball_placement_play_test.py` does not have any validation. Validation needs to be added for this test (lines 84 and 88). A good starting point would be looking at `ball_placement_play_test.cpp`

### Acceptance criteria

- [ ] Always validation for ball placement
- [ ] Eventually validation for ball placement

### PR DESCRIPTION
**Pull request by jameschuang2002**
### Resolved Issues
resolve #2780

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by jameschuang2002** [comment]
This branch should be merged after #3001 given that the changes were incorporated

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 1]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+# send a free kick command
+# simulated_test_runner.gamecontroller.send_ci_input(
+#     gc_command=Command.Type.DIRECT_FREE_BLUE, team=Team.BLUE
+# )
+
+# # Drop Ball Always Validation
+# drop_ball_always_validation_sequence_set = [
+#     [
+#         BallAlwaysStaysInRegion(
+#             regions=[tbots.Circle(ball_final_pos, 0.15)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
#             regions=[tbots_cpp.Circle(ball_final_pos, 0.15)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 2]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+# ]
+
+# # Drop Ball Eventually Validation
+# # Direct free kick after ball placement, the robot must be 0.05 away from the ball after the placement
+# # See detailed rules here: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement
+# drop_ball_eventually_validation_sequence_set = [
+#     [
+#         # Ball should arrive within 0.15m of placement point
+#         BallEventuallyStopsInRegion(
+#             regions=[tbots.Circle(ball_final_pos, 0.15)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
#             regions=[tbots_cpp.Circle(ball_final_pos, 0.15)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 3]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+#           #Robot should exit from ball at least 0.05m
+#         RobotEventuallyExitsRegion(
+#             regions=[tbots.Circle(ball_final_pos, 0.2)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
#             regions=[tbots_cpp.Circle(ball_final_pos, 0.2)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 4]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+    # send a non free kick command after the ball ball placement command
+    # simulated_test_runner.gamecontroller.send_ci_input(
+    #     gc_command=Command.Type.FORCE_START, team=Team.BLUE
+    # )
+
+    # # Drop Ball Always Validation
+    # drop_ball_always_validation_sequence_set = [
+    #     [
+    #         BallAlwaysStaysInRegion(
+    #             regions=[tbots.Circle(ball_final_pos, 0.1)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
    #             regions=[tbots_cpp.Circle(ball_final_pos, 0.1)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 5]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+    # ]
+
+    # # Drop Ball Eventually Validation
+    # # Non free kick after ball placement, the robot must be 0.5 away from the ball after the placement
+    # # See detailed rules here: https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement
+    # drop_ball_eventually_validation_sequence_set = [
+    #     [
+    #         # Ball should arrive within 0.15m of placement point
+    #         BallEventuallyStopsInRegion(
+    #             regions=[tbots.Circle(ball_final_pos, 0.15)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
    #             regions=[tbots_cpp.Circle(ball_final_pos, 0.15)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 6]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+    #           # Robot has to be 0.5 m away from the ball
+    #         RobotEventuallyExitsRegion(
+    #             regions=[tbots.Circle(ball_final_pos, 0.65)]),
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]:
```suggestion
    #             regions=[tbots_cpp.Circle(ball_final_pos, 0.65)]),
```

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 7]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+                RobotEventuallyEntersRegion(
+                    regions=[tbots_cpp.Circle(ball_final_pos, 0.15)]
                 ),
             ]
         ],
         test_timeout_s=[20],
     )
 
 
+# TODO (#2797): uncomment tests below to verify robots actually drop ball and exit region
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]: can you make these commented lines in line with the test above?

File: src/software/ai/hl/stp/play/ball_placement_play_test.py [hunk 8]
```diff
@@ -92,24 +93,198 @@ def setup(run_enemy_ai):
...
+# send a free kick command
+# simulated_test_runner.gamecontroller.send_ci_input(
+#     gc_command=Command.Type.DIRECT_FREE_BLUE, team=Team.BLUE
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/ball_placement_play_test.py]: this line is problematic for some reason. Can you update this file and call the gc_command types differently?

```
from proto.ssl_gc_referee_message_pb2 import Referee
```

Then modify the calls to the referee command as follows:
```
simulated_test_runner.gamecontroller.send_ci_input(
    gc_command=Referee.Command.DIRECT_FREE_BLUE, team=Team.BLUE

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 7 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: @jameschuang2002 bump

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 8 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: @jameschuang2002 bump

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 1 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: @jameschuang2002 bump

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 2 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: @jameschuang2002 bump

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 3 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: @jameschuang2002 bump

**Comment by nimazareian** [comment]
@jameschuang2002 could you pull master and make sure CI is passing for this PR?

Also, should we remove the ball placement cpp tests? cc: @itsarune

**Comment by itsarune** [comment]
Yes it's true, we can delete the `ball_placement_play_test.cpp` test with this pytest

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 7 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: I am not sure what this is meaning. Is it to move the code segment into where it should replace? I think that might make it more difficult to read.

File: src/software/ai/hl/stp/play/ball_placement_play_test.py (hunk 8 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/ball_placement_play_test.py]: resolved after pull from main

File: src/software/simulated_tests/robot_enters_region.py [hunk 9]
```diff
@@ -12,7 +12,7 @@ class RobotEntersRegion(Validation):
 
     """Checks if a Robot enters any of the provided regions."""
 
-    def __init__(self, regions=None):
+    def __init__(self, regions=None, num_robots=1):
```
Comment (nimazareian) [inline src/software/simulated_tests/robot_enters_region.py:15]: nit: This variable doesn't seem to be used

File: src/software/simulated_tests/ball_moves_forward.py [hunk 10]
```diff
@@ -57,7 +57,7 @@ def __repr__(self):
 
 (
     BallEventuallyMovesForward,
-    BallEventuallyStopsMovingForward,
+    BallStopsMovingForward,
```
Comment (nimazareian) [inline src/software/simulated_tests/ball_moves_forward.py:60]: is this change actually being used?

**Comment by itsarune** [comment]
merged in via #3046

**Pull request closed without merging**